### DISCUSSION TITLE
Property 'pick' does not exist on type 'ZodUnion'

### DISCUSSION THREAD

**Discussion by baptisteArno** [colinhacks/zod#2164] [Q&A]
My schema is a union of multiple other schemas.

Even though the other schemas share the same shape, I can't call the `pick` method on the schema.
`Property 'pick' does not exist on type 'ZodUnion'`

Is there any work around for that?

EDIT:

I found about `merge` instead of `and`. Is there something similar for unions?

**Comment by JacobWeisenburger** [comment]
Is this what you are looking for?

> Even though the other schemas share the same shape, I can't call the `pick` method on the schema.
> `Property 'pick' does not exist on type 'ZodUnion'`

```ts
const schema = z.union( [
    z.object( {
        foo: z.string(),
        bar: z.string(),
    } ),
    z.object( {
        foo: z.string(),
    } ),
] )

const hasFoo = schema._def.options[ 0 ].pick( { foo: true } )
type HasFoo = z.infer<typeof hasFoo>
// type HasFoo = {
//     foo: string
// }
```

> I found about `merge` instead of `and`. Is there something similar for unions?

```ts
const baseObj = z.object( {
    foo: z.string(),
} )
const union = z.union( [
    baseObj.extend( { bar: z.number() } ),
    baseObj.extend( { baz: z.boolean() } ),
] )
type Union = z.infer<typeof union>
// type Union = {
//     foo: string
//     bar: number
// } | {
//     foo: string
//     baz: boolean
// }
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by baptisteArno** [reply]
Thank you for the help! But I think that's not it!

I'm trying to reproduce this behavior:

```ts
type Car =  {
    type : "car"
}

type Plane =  {
    type : "plane"
}

type Vehicule = Car | Plane

type Type = Pick<Vehicule, 'type'>
```

This won't work:

```ts
const car = z.object({
  type: z.literal('car')
})

const plane = z.object({
  type: z.literal('plane')
})

const vehicule = z.union([car, plane])

const type = vehicule.pick({type: true})
```

**Comment by JacobWeisenburger** [comment]
Is this what you are looking for?

```ts
const type = z.enum( [ 'car', 'plane' ] )
const car = z.object( { type: type.extract( [ 'car' ] ) } )
const plane = z.object( { type: type.extract( [ 'plane' ] ) } )
const vehicle = z.union( [ car, plane ] )
type Vehicle = z.infer<typeof vehicle>
// type Vehicle = { type: 'car' } | { type: 'plane' }
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by baptisteArno** [reply]
That doesn't solve the example I need:

This works fine

```ts
const car = z.object({
  type: z.literal('car')
  name: z.string()
})

const plane = z.object({
  type: z.literal('plane')
  name: z.string()
})

const vehicule = z.union([car, plane])
```

But then I cannot just pick type or name for example on the union

**Reply by Alexandregirbal** [reply]
The `.union` considers the car and plane schemas as **totally different** objects. To solve your issue, you can think "reversed" and create a parent object (vehicule) that will either be extended by children (if you have different fields) or just use an enum like so:
```typescript
const vehiculeTypes = z.enum(["car", "plane"]);

const vehicule = z.object({
    name: z.string(),
    type: vehiculeTypes,
});

const type = vehicule.pick({ name: true });
````
